pennylane/gradients/pulse_generator_gradient.py

```diff
@@ -0,0 +1,578 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for computing the pulse generator
+parameter-shift gradient of pulse sequences in a qubit-based quantum tape.
+"""
+from collections.abc import Sequence
+import numpy as np
+
+import pennylane as qml
+
+from pennylane.pulse import ParametrizedEvolution
+from pennylane.ops.qubit.special_unitary import pauli_basis_strings, _pauli_decompose
+
+from .parameter_shift import _make_zero_rep
+from .pulse_gradient import _assert_has_jax
+from .gradient_transform import (
+    _all_zero_grad,
+    assert_active_return,
+    assert_no_state_returns,
+    assert_no_variance,
+    choose_grad_methods,
+    gradient_analysis_and_validation,
+    gradient_transform,
+    _no_trainable_grad,
+    reorder_grads,
+)
+
+try:
+    import jax
+except ImportError:
+    # Handling the case where JAX is not installed is done via _assert_has_jax
+    pass
+
+
+def _one_parameter_generators(op):
+    r"""Compute the effective generators of one-parameter groups that reproduce
+    the partial derivatives of a parameterized evolution.
+
+    Args:
+        op (`~.ParametrizedEvolution`): Parametrized evolution for which to compute the generator
+
+    Returns:
+        tuple[tensor_like]: The generators for one-parameter groups that reproduce the
+        partial derivatives of the parametrized evolution.
+        The ``k``\ th entry of the returned ``tuple`` has the shape ``(2**N, 2**N, *par_shape)``
+        where ``N`` is the number of qubits the evolution acts on and ``par_shape`` is the
+        shape of the ``k``\ th parameter of the evolution.
+
+    See the documentation of pulse_generator for more details and a mathematical derivation.
+    """
+
+    def _compute_matrix(*args):
+        return op(*args, t=op.t, **op.odeint_kwargs).matrix()
+
+    data = [(1.0 + 0.0j) * d for d in op.data]
+    # These lines compute the Jacobian of compute_matrix every time -> to be optimized
+    jac = jax.jacobian(_compute_matrix, argnums=0, holomorphic=True)(data)
```

Comment:
I've found that it is faster to split the _compute_matrix output into real and imaginary part of the matrix, compute the respective jacobians with real parameters, and recombine them appropriately. Could be worth a shot here! jacobian with holomorphic=True is slightly slower (probably due to having to use all-complex parameters.

Reply:
Oh, really? :o This surprises me because one has to trace the matrix computation twice. But then again, this is probably cheap compared to actually doing all the FLOPs :)